### PR TITLE
created fuid search

### PR DESCRIPTION
**Pull request by tala-abuadas**

### CODE REVIEW

File: src/foam/dao/RequestResponseClientDAO.js [hunk 1]
```diff
@@ -106,7 +106,7 @@ Suitable for usage against backends that don't support listen(), such as plain H
           this.on.reset.pub();
           return true;
         }
-        if ( ! foam.dao.DAO.PURGE_CMD === obj ) {
+        if ( !(foam.dao.DAO.PURGE_CMD === obj) ) {
```
Comment (sarthak-marwaha) [inline src/foam/dao/RequestResponseClientDAO.js]:
```suggestion
        if ( ! ( foam.dao.DAO.PURGE_CMD === obj ) ) {
```

Comment (kgrgreer) [reply src/foam/dao/RequestResponseClientDAO.js]: of better: !==

File: src/foam/u2/view/GlobalFuidSearch.js [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+  package: 'foam.u2.view',
+  name: 'GlobalFuidSearch',
+  extends: 'foam.u2.view.SuggestedTextField',
+
+  requires: [
+    'foam.u2.FUIDAutocompleter',
```
Comment (kgrgreer) [inline src/foam/u2/view/GlobalFuidSearch.js]: sort requires

File: src/foam/u2/view/GlobalFuidSearch.js [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+      of: 'foam.u2.FUIDAutocompleter',
+      name: 'autocompleter',
+      factory: function() {
+        return this.FUIDAutocompleter.create();
+      }
+    },
+    {
+      class: 'String',
+      name: 'title',
+      factory: function() {
```
Comment (kgrgreer) [inline src/foam/u2/view/GlobalFuidSearch.js]: Why is this needed? The default value: for Strings is '' already?

File: src/foam/dao/FUIDDAO.js [hunk 4]
```diff
@@ -128,6 +136,14 @@ foam.CLASS({
...
         return getDelegate().put_(x, obj);
       `
+    },
+    {
+      name: 'cmd_',
+      args: 'Context x, Object obj',
+      javaCode:
+      `
+        return getSalt();
+      `
```
Comment (lchanmann) [inline src/foam/dao/FUIDDAO.js]: Don't think it needs to expose the salt, all fuid keys are in fuidKeyDAO.

Comment (sarthak-marwaha) [reply src/foam/dao/FUIDDAO.js]: FUID keys are only on lowest level daos, we need to find keys for all daos that can be accessed through the menuDAO

File: src/foam/u2/view/GlobalFuidSearch.js (hunk 3 above)
Comment (tala-abuadas) [reply src/foam/u2/view/GlobalFuidSearch.js]: im overriding parent class factory

Comment (kgrgreer) [reply src/foam/u2/view/GlobalFuidSearch.js]: Ok, then add a comment so that someone doesn't just delete this in the future thinking its redundant. Thx

File: src/foam/u2/view/GlobalFuidSearch.js [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+      of: 'foam.u2.FUIDAutocompleter',
+      name: 'autocompleter',
+      factory: function() {
+        return this.FUIDAutocompleter.create();
+      }
+    },
+    {
+      class: 'String',
+      name: 'title',
+      value: ''
```
Comment (kgrgreer) [inline src/foam/u2/view/GlobalFuidSearch.js:60]: add comment

File: src/foam/u2/view/GlobalFuidSearch.js [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+    },
+    {
+      class: 'String',
+      name: 'emptyTitle',
+      value: ''
```
Comment (kgrgreer) [inline src/foam/u2/view/GlobalFuidSearch.js:66]: add comment

File: src/foam/dao/FUIDDAO.js (hunk 4 above)
Comment (lchanmann) [reply src/foam/dao/FUIDDAO.js]: so for when searching a business id, the view will show one item with userDAO and another item with the businessDAO?

also do we still want to fall through if it's not the SALT_CMD?

Comment (sarthak-marwaha) [reply src/foam/dao/FUIDDAO.js]: Yes it will show both DAOs and lets the user pick.

Comment (sarthak-marwaha) [reply src/foam/dao/FUIDDAO.js]: right it should check for SALT_CMD and otherwise call SUPER()

Comment (lchanmann) [reply src/foam/dao/FUIDDAO.js]: > Yes it will show both DAOs and lets the user pick.

Then should it also consider the predicate on the menu eg. userDAO only shows User objects and businessDAO only shows Business objects?

Comment (sarthak-marwaha) [reply src/foam/dao/FUIDDAO.js]: It does a find before showing the option to open it in the menu so if the find returns undefined it will hide the menu. L:77 of GlobalFUIDSearch.js

File: src/foam/u2/FUIDAutocompleter.js [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+  listeners: [
+    {
+      name: 'onUpdate',
+      isFramed: true,
+      code: async function onUpdate() {
+        var res = this.partial ? await this.globalSearchService.searchById(this, this.partial) : [];
+        var temp = [];
+        for ( var [daoKey, obj1] of Object.entries(res) ) {
+          temp.push({ daoKey: daoKey, obj1: obj1 });
```
Comment (lchanmann) [inline src/foam/u2/FUIDAutocompleter.js]: could we rename the obj1 to data?
```suggestion
          temp.push({ daoKey: daoKey, data: data });
```

File: src/foam/dao/FUIDDAO.js (hunk 4 above)
Comment (lchanmann) [reply src/foam/dao/FUIDDAO.js]: ok, I see. Please add the check for SALT_CMD and otherwise call SUPER(). and can we have a demo of this tomorrow?

File: src/foam/dao/FUIDDAO.js [hunk 8]
```diff
@@ -128,6 +136,19 @@ foam.CLASS({
...
+      name: 'cmd_',
+      args: 'Context x, Object obj',
+      javaCode:
+      `
+        if ( FUIDDAO.SALT_CMD.equals(obj) ) {
+          return getSalt();
+        }
+        else {
+          return getDelegate().cmd_(x, obj);
+        }
```
Comment (lchanmann) [inline src/foam/dao/FUIDDAO.js]:
```suggestion
        return getDelegate().cmd_(x, obj);
```

**Review by lchanmann — APPROVED** [review]

**Review by sarthak-marwaha — APPROVED** [review]
